HDDS-11938. [Website v2] [Docs] [Quick Start] Try Ozone With Kubernetes (#116)
apache/ozone-site: docs/02-quick-start/01-installation/02-kubernetes.md

s3g-public   NodePort    10.97.133.137   <none>        9878:31880/TCP   27s
scm          ClusterIP   None            <none>        9876/TCP         27s
scm-public   NodePort    10.105.231.28   <none>        9876:32171/TCP   27s
```

Minikube contains a convenience command to access any of the NodePort services:

``` bash
minikube service s3g-public
Opening kubernetes service default/s3g-public in default browser...
```

## Hosted Kubernetes Cluster

**Requirements:**

* Working Kubernetes cluster (LoadBalancer, PersistentVolume are not required)
* kubectl

As *apache/ozone* Docker images are available from Docker Hub, the deployment process is very similar to Minikube deployment. The only big difference is that we have dedicated set of k8s files for hosted clusters (for example we can use one Datanode per host)

### Deploy to Kubernetes

`kubernetes/examples` folder of the Ozone distribution contains Kubernetes deployment resource files for multiple use cases.

To deploy to a hosted cluster use the Ozone subdirectory:

``` bash
cd kubernetes/examples/ozone
kubectl apply -f .
```

And you can check the results with

``` bash
kubectl get pod
Access the services
```

> 

Reviewer: This is also supposed to be a heading:

```
### Access the services
```

Author: If I add the heading, it would result in the MD024/no-duplicate-heading error at /root/project/ozone-site/docs/02-quick-start/01-installation/02-kubernetes.md:100 under '### Access the services'. Should I simply drop the heading?

Reviewer: Hmm, good point.

Possible solutions:
- split the page, as in the existing website
- come up with unique headings
- remove these subsection headings
- is there some flag to allow duplicate headings?

@errose28 what's your take?

Reviewer 2: Ah this was a miss when I set up the CI. We only want the duplicate header check for siblings, which is not the case here. See https://github.com/DavidAnson/markdownlint/blob/v0.35.0/doc/md024.md and please set `siblings_only=true` in `markdownlint.yaml` as part of this PR to fix this issue. Then these headings should be allowed.

---
sidebar_label: Kubernetes
---
<!---
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

> -->

Reviewer: I think license is included in the original docs in `apache/ozone` since they are part of the official source release.  But the website is not "released", so we can remove the license header.

# Try Ozone With Kubernetes

## Minikube

**Requirements:**

* Working minikube setup

> * kubectl

Reviewer 2: Please use `-` for unordered lists since it is consistent with the existing pages in the new site. This was supposed to be enforced by markdownlint but I made a mistake in the config. I'll push a fix for the markdownlint config soon.

> 

Reviewer 2: ```suggestion
### Deploy the Services
```

> `kubernetes/examples` folder of the Ozone distribution contains Kubernetes deployment resource files for multiple use cases. By default the Kubernetes resource files are configured to use *apache/ozone* image from the Docker Hub.

Reviewer 2: ```suggestion
The `kubernetes/examples` directory of the Ozone distribution contains Kubernetes deployment resource files for multiple use cases. By default the Kubernetes resource files are configured to use the [apache/ozone](https://hub.docker.com/r/apache/ozone) image from Docker Hub.
```

To deploy it to minikube, use the minikube configuration set:

``` bash
cd kubernetes/examples/minikube
kubectl apply -k .
```

And you can check the results with

``` bash
kubectl get pod
```

> Note: the `kubernetes/examples/minikube` resource set is optimized for minikube usage:

Reviewer 2: For callouts like this we should docusaurus admonitions as described in the [contributing guide](https://github.com/apache/ozone-site/blob/HDDS-9225-website-v2/CONTRIBUTING.md#page-format).

* You can have multiple Datanodes even if you have only one host (in a real production cluster usually you need one Datanode per physical host)

> * The services are published with node port

Reviewer 2: Can you elaborate on what this means?

Author: Node port is an opening port that allowing externel traffic to access the service.
The detailed explaination could be found [here](https://kubernetes.io/docs/concepts/services-networking/service/)

### Access the services

> Now you can access any of the services. For each web endpoint an additional NodeType service is defined in the minikube k8s resource set. NodeType services are available via a generated port of any of the host nodes:

Reviewer 2: ```suggestion
Now you can access any of the services. For each web endpoint an additional NodeType service is defined in the minikube k8s resource set. NodeType services are available via a generated port on any of the host nodes:
```

``` bash
kubectl get svc
NAME         TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)          AGE
datanode     ClusterIP   None            <none>        <none>           27s
kubernetes   ClusterIP   10.96.0.1       <none>        443/TCP          118m
om           ClusterIP   None            <none>        9874/TCP         27s
om-public    NodePort    10.108.48.148   <none>        9874:32649/TCP   27s
s3g          ClusterIP   None            <none>        9878/TCP         27s
s3g-public   NodePort    10.97.133.137   <none>        9878:31880/TCP   27s
scm          ClusterIP   None            <none>        9876/TCP         27s
scm-public   NodePort    10.105.231.28   <none>        9876:32171/TCP   27s
```

> Minikube contains a convenience command to access any of the NodePort services:

Reviewer 2: ```suggestion
Minikube provides a convenience command to access any of the NodePort services:
```

``` bash
minikube service s3g-public

> Opening kubernetes service default/s3g-public in default browser...

Reviewer 2: I haven't tried this, but what does this display? Is it the s3g web UI?

Author: Yes, it will provide a URL to access a website.
![image](https://github.com/user-attachments/assets/a3137146-07e1-40d4-a9bb-3b6baa3fa35d)

```

> ## Hosted Kubernetes Cluster

Reviewer 2: The formatting suggestions above also apply to this section

**Requirements:**

* Working Kubernetes cluster (LoadBalancer, PersistentVolume are not required)
* kubectl

As *apache/ozone* Docker images are available from Docker Hub, the deployment process is very similar to Minikube deployment. The only big difference is that we have dedicated set of k8s files for hosted clusters (for example we can use one Datanode per host)

### Deploy to Kubernetes

`kubernetes/examples` folder of the Ozone distribution contains Kubernetes deployment resource files for multiple use cases.

To deploy to a hosted cluster use the Ozone subdirectory:

``` bash
cd kubernetes/examples/ozone
kubectl apply -k .
```

And you can check the results with

``` bash
kubectl get pod
```

### Access the services

Now you can access any of the services. By default the services are not published but you can access them with port-forward rules.

``` bash
kubectl port-forward s3g-0 9878:9878
kubectl port-forward scm-0 9876:9876

> ```

Reviewer 2: At some point here the we should run `ozone version` in a pod as part of the test that things were set up correctly. More thorough testing will be done in the following "Reading and Writing Data" section.

---
<!---
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->

# Try Ozone With Kubernetes

**Requirements:**

* Working kubernetes cluster (LoadBalancer, PersistentVolume are not required)
* kubectl

As the *apache/ozone* Docker images are available from the dockerhub the deployment process is very similar to Minikube deployment. The only big difference is that we have dedicated set of k8s files for hosted clusters (for example we can use one Datanode per host)

> Deploy to kubernetes

Reviewer: This is supposed to be a heading (was wrong in the existing docs).

```suggestion
### Deploy to Kubernetes
```

(Suggesting 3rd level heading, assuming Minikube and "hosted cluster" content would be 2nd level, and this is under "hosted cluster".)
